pacman/operations/routing_info_allocator_algorithms/zoned_routing_info_allocator.py

```diff
@@ -16,154 +16,308 @@
+        # For each App vertex / Partition name zone keep track of the number of
+        # bites required for the mask for each machine vertex
+        "__atom_bits_per_app_part",
+        # maximum number of bites to represent the keys and masks
+        # for a single app vertex / partition name zone
+        "__n_bits_atoms_and_mac",
+        # Maximum number pf bit to represent the machine assuming global
+        "__n_bits_machine",
+        # Maximum number pf bit to represent the machine assuming global
+        "__n_bits_atoms",
+        # Flag to say operating in flexible mode
+        "__flexible",
+        # List of (key, n_keys) needed for fixed
+        "__fixed_keys",
+        # Map of partition to fixed_+key_and_mask
+        "__fixed_partitions",
+        # Set of app_part indexes used by fixed
+        "__fixed_used"
     ]
-    MAX_PARTITIONS_SUPPORTED = 1
     # pylint: disable=attribute-defined-outside-init
 
-    def __call__(self, application_graph, machine_graph, n_keys_map):
+    def __call__(self, machine_graph, n_keys_map, flexible):
         """
-        :param ApplicationGraph application_graph:
         :param MachineGraph machine_graph:
+            The machine graph to allocate the routing info for
         :param AbstractMachinePartitionNKeysMap n_keys_map:
-        :rtype: tuple(RoutingInfo, dict(ApplicationVertex, BaseKeyAndMask))
+            A map between the edges and the number of keys required by the
+            edges
+        :param bool flexible: Dettermines if flexible can be use.
+            If False global settings will be attempted
         :raise PacmanRouteInfoAllocationException:
         """
         # check that this algorithm supports the constraints put onto the
         # partitions
-        self.__application_graph = application_graph
         self.__machine_graph = machine_graph
         self.__n_keys_map = n_keys_map
+        self.__n_bits_atoms_and_mac = 0
+        self.__n_bits_machine = 0
+        self.__n_bits_atoms = 0
+        self.__atom_bits_per_app_part = dict()
+        self.__flexible = flexible
+        self.__fixed_keys = list()
+        self.__fixed_partitions = dict()
+        self.__fixed_used = set()
 
         check_algorithm_can_support_constraints(
             constrained_vertices=machine_graph.outgoing_edge_partitions,
-            supported_constraints=[ContiguousKeyRangeContraint],
+            supported_constraints=[
+                ContiguousKeyRangeContraint, FixedKeyAndMaskConstraint],
             abstract_constraint_type=AbstractKeyAllocatorConstraint)
 
-        max_app_keys_bits, key_bits_per_app = self._calculate_zones()
+        self.__find_fixed()
+        self.__calculate_zones()
+        self.__check_zones()
+        return self.__allocate()
 
-        return self._simple_allocate(max_app_keys_bits, key_bits_per_app)
+    def __find_fixed(self):
+        by_app_and_partition_name = \
+            self.__machine_graph.multicast_partitions
```

Comment:
As mentioned elsewhere, would this be better as simply "machine_partitions"?

Reply:
There are two levels of map
Key level 1 app_id
key Level 2 partitoin_name
Value  list of vertices

Reply:
The issue is more that the name doesn't say *what* is by app and partition name.... just that something is!  You have to go back to this line to know what it is... Maybe then m_partition_map or something...

Reply:
If you want to go all Alan here, you could call it machine_partitions_by_app_and_partition_name... I don't really recommend this though!

Reply:
rename attempted